### PR TITLE
[PLS Refactor] Linear solver information operators

### PR DESCRIPTION
**Pull request by JonathanWenger**
# In a Nutshell
Implements information operators for the probabilistic linear solver.

# Detailed Description
This PR implements functionality described by a information operator, which takes a `ProbabilisticLinearSolverState` including a current action and extracts information about a linear system. The following classes are added:
- `LinearSolverInfoOp` abstract base class
- `MatVecInfoOp`: obtain information about the linear system via matrix-vector multiplication
- `ProjResidualInfoOp`: observe the linear system by projecting the residual onto an arbitrary vector

# Related Issues and PRs
First part of #51 and supersedes part of #283.

### CODE REVIEW

File: tests/test_linalg/test_solvers/cases/states.py [hunk 1]
```diff
@@ -1,25 +1,46 @@
...
-def case_linear_solver_state(
+# Prior
+prior = linalg.solvers.beliefs.LinearSystemBelief(
+    A=randvars.Constant(linsys.A),
+    Ainv=None,
+    x=randvars.Normal(
+        mean=np.zeros(linsys.A.shape[1]), cov=linops.Identity(shape=linsys.A.shape)
+    ),
+    b=randvars.Constant(linsys.b),
+)
```
Comment (marvinpfoertner) [inline tests/test_linalg/test_solvers/cases/states.py:23]: Do we want to make fixtures out of these?

**Review by marvinpfoertner — APPROVED** [review]
Very nice, I like the amount of `pytest_cases` features used.
We should definitely integrate these much more throughout the library :)

File: src/probnum/linalg/solvers/information_ops/_linear_solver_info_op.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
+"""Base class for linear solver information operators."""
+import abc
+
+import numpy as np
+
+import probnum  # pylint: disable="unused-import"
```
[the remark this answers is not in the document]
Comment (JonathanWenger) [reply src/probnum/linalg/solvers/information_ops/_linear_solver_info_op.py:6]: Needed for the type hint forward reference ```"probnum.linalg.solvers.ProbabilisticLinearSolverState"```.

File: src/probnum/linalg/solvers/information_ops/__init__.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
+"""Information operators of probabilistic linear solvers."""
```
[the remark this answers is not in the document]
Comment (JonathanWenger) [reply src/probnum/linalg/solvers/information_ops/__init__.py]: I added some more explanation for linear solvers here.

File: tests/test_linalg/test_solvers/cases/states.py (hunk 1 above)
Comment (JonathanWenger) [reply tests/test_linalg/test_solvers/cases/states.py:23]: I'll leave them as cases for now and ina future PR go over the test structure again to see whether it makes sense to have fixtures for those objects that are not compositional (and thus depend on other cases), or whether we just build up everything from cases.

**Pull request merged**